### ISSUE TITLE
Shuffling after every insert set?

### ISSUE THREAD

**Issue by brurucy** [tidwall/btree-benchmark#1]
Hi,

I noticed that you sort, or shuffle, the integer set after every measurement.

Why so?

**Comment by tidwall** [comment]
Sorting every time is probably not needed. Likely only once before the first sequential benchmarks is enough.

I'm shuffling before executing each random group of operations because I want to ensure that all entries in memory are different, in hopes to minimize cpu optimizations like caching and such.

**Comment by brurucy** [comment]
But isn't the main point of B-Tree to take advantage of the cache? 

I believe you are also not flushing the instruction set cache.

Would it be possible to just create an alternative benchmark, without the intermittent shuffling, and with higher values? maybe up to 10 or 100 million?

**Comment by tidwall** [comment]
> But isn't the main point of B-Tree to take advantage of the cache?

I'm not sure that that is specifically the main point of a B-tree, but a B-tree certainly does a good job of taking advantage of modern cpu caches. Beside, the 'random' benchmarks only care about the contiguous array of ints that are inserted into the B-tree. I want to shuffle that memory prior to inserting into the B-tree. The cache as it relates to the B-tree itself is just a property of the implementation and cpu. 

> I believe you are also not flushing the instruction set cache.

Flushing the cache is not a goal of the benchmark.

> Would it be possible to just create an alternative benchmark, without the intermittent shuffling, and with higher values? maybe up to 10 or 100 million?

Absolutely! There's no hard coded upper limit.  I encourage you to fork the project and experiment with different strategies. 

If you are researching for academic purposes it may also be worthwhile to look at the https://github.com/tidwall/btree.c too, which is a C implementation with the same logic. 

And the C++ one by Google is pretty excellent too. https://code.google.com/archive/p/cpp-btree/

**Comment by brurucy** [comment]
Thanks for the response.

I have a question, how much of a slowdown does one get by using interfaces as key values over regular native datatypes? It surely seems to be the most optimal way of doing it, given that there's no generics yet, but I wonder how big is the performance impact.

**Comment by tidwall** [comment]
Using native datatypes would be more optimal. Generics could allow for custom data types to be stored directly in the B-tree nodes, instead of using an interface{} (which is a 16-byte [type-signature,pointer] to the data on 64-bit arch).

> how much of a slowdown does one get by using interfaces as key values over regular native datatypes?

Depends on what is being stored in the B-tree. If you are storing small, short-lived objects then the benefit could be big because it may be possible to avoid the extra overhead of the 16-byte interface{} container and the extra allocation and pointer in the heap.

If you are storing large or long-lived objects, the kind that might be are shared in across other data structures or used by goroutines, then there's a good chance that you are already storing the objects as its own pointer on heap anyhow, and the benefit may be less.

For example, in the [Tile38](https://github.com/tidwall/tile38) project I use this B-tree implementation to store all my objects. I also store all the same objects in an R-tree. There's often millions of these objects, each ranging from 50 bytes up to a few KB, and they must be pointers on the heap. In this scenario, generics could help a little, but it's probably going to be nominal.

The other thing generics bring is the potential of inlining the compare function.  The C++ implementation that I mentioned in my previous comment can support that.

It's likely, at some point, that I'll upgrade this B-tree to support generics.
